calliope/core/util/observed_dict.py

```diff
@@ -0,0 +1,73 @@
+
+"""
+Copyright (C) 2013-2019 Calliope contributors listed in AUTHORS.
+Licensed under the Apache 2.0 License (see LICENSE file).
+
+"""
+
+from collections import Mapping
+from calliope.core.attrdict import AttrDict
+
+
+class ObservedDict(dict):
```

Comment:
Could or should these be subclasses of AttrDict or does that just mess things up completely?

Reply:
It messes things up completely, since `.notify` becomes a key as well as a method. I would have preferred it to be a subclass of AttrDict, for the ease of going to and from YAML, but couldn't see a way of doing it.

Reply:
Or, perhaps more rightly, `notify` could be overridden by a key being assigned with the same name. Do you protect certain attributes/methods in the current implementation of AttrDict?

Reply:
Had another go at reimplementing this and managed to remove `on_changed` as an issue, but `observer` and 'name' still get added as keys in the dictionary. Unlike `on_changed`, they cannot be passed as explicit arguments to `notify` (I don't think)